### ISSUE TITLE
Interactions between main routine and sub-routine

### ISSUE THREAD

**Issue by wenyaowu** [cjlee112/socraticqs2#22]
In the random trial stage, we need a mechanism to go back and forward between the main routine (Start-Pre test-Treatments-Post test-end) and subroutines (which is the FSM of pre/post test and Treatments).

We can use FSMStack.push() to load the selected subroutine(unit), and as the last state of subroutine in the FSMStack is popped, send 'END' event to the main-routine then stepping in to the next state of the main-routine until the main-routine step into the 'END' state.

**Comment by cjlee112** [comment]
checking the latest code (fsm branch) I see that the event names that are sent on sub-FSM closure are

* "pop": sub-FSM completed normally, i.e. reached its END state;
* "abort": user cancelled the sub-FSM (using the Activity Center view, aka node_status()).

Reflecting a bit, I think "pop" may be better (clearer, more explicit about exactly what's happening) than "done".  "pop" articulates that a sub-FSM popped, whereas "done" sounds like the current FSM is done.  What do you think?  I'm not sure I have a preference between "abort" vs. "cancel".

Perhaps another way of thinking about this is by analogy to Python "return" vs. "raise".  "return" represents normal completion of a subroutine, returning a result.  "raise" represents abnormal failure to complete a subroutine.  So perhaps we should call the normal completion event "return" (or "returned"?).  And perhaps we should call the abort event something like "exceptCancel"?  We could model this to behave like Python try-except; that is, if a node doesn't provide an "exceptCancel" edge, then the exception just propagates up the stack, i.e. this FSM is also popped and "exceptCancel" passed to its parent FSM if any.  That provides a reasonable default behavior for cancelation, so that we wouldn't have to provide explicit "exceptCancel" edges unless we wanted to "catch" the exception and handle it in a special way.  What do you think?

**Comment by cjlee112** [comment]
If no one else has an opinion on this, I think I'll go ahead with the "return" / "exceptCancel" proposal outlined in the previous paragraph.

**Comment by wenyaowu** [comment]
Yeah, since we are using Stack, I think 'pop' will be better choice.
Also, return/ raise exception is a good proposal, I've seen people using this on checking HttpRequest/ Response. It will be easier for us to examine if the subroutine send back good or bad resoponse.

**Comment by wenyaowu** [comment]
![10965648_832954060084761_57540123_n](https://cloud.githubusercontent.com/assets/6733162/6195807/7d7f3724-b386-11e4-90f9-2a9f79704357.jpg)
